### PR TITLE
fix(controller): Use node.Name instead of node.DisplayName for onExit nodes

### PR DESCRIPTION
**Pull request by simster7**
Split from https://github.com/argoproj/argo-workflows/pull/5478

### CODE REVIEW

File: workflow/controller/steps.go [hunk 1]
```diff
@@ -260,7 +260,7 @@ func (woc *wfOperationCtx) executeStepGroup(ctx context.Context, stepGroup []wfv
 		if !childNode.Fulfilled() {
 			completed = false
 		} else if childNode.Completed() {
-			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, step.Name, childNode.Name, stepsCtx.boundaryID, stepsCtx.tmplCtx)
+			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, childNode.Name, stepsCtx.boundaryID, stepsCtx.tmplCtx)
```
Comment (simster7) [inline workflow/controller/steps.go:263]: `parentDisplayName` is non unique, causing nodes with the same display name (e.g. when using `withParam`) to create onExit nodes of the same name. After the first onExit node is created, subsequent onExit nodes created by different parent nodes (but with the same display name) will fail creation as they already exist. Use `parentNodeName` which is always unique instead.

**Review by terrytangyuan — APPROVED** [review]

**Comment by alexec** [comment]
what happens to in-flight workflows? can do they end up with 2x on-exit nodes?

**Comment by simster7** [comment]
> what happens to in-flight workflows at the time of upgrade? can do they end up with 2x on-exit nodes?

Yes, there is a window of time where onExit nodes could get duplicated in the event of an upgrade while a cluster is running. note this window is only from the time an onExit node with the old name is created until the time its boundary node (Step Group, DAG) completes. Once the boundary node completes, we won't recurse down to the onExit node again.

I don't believe we make guarantees on the consistency of a running workflow if there is a version upgrade while it's running.

**Comment by alexec** [comment]
I don't think many users expect an upgrade to typically break workflows. Is there an alternative you can see?

**Comment by simster7** [comment]
> I don't think many users expect an upgrade to typically break workflows. Is there an alternative you can see?

@alexec I don't believe so. This is a narrow issue: are we using the right name or not? We are currently not using the right name, which means there are bugs in the code (see https://github.com/argoproj/argo-workflows/pull/5486#discussion_r599004491).

If it's critical that no running workflows are broken during the upgrade, I could introduce some code that:

* first checks if a node exists with the old name
* if it does, proceed with the old name
* if it doesn't, proceed with the new name

We'd have to keep this scaffold code in the codebase for a couple of versions to ensure that users upgrading and skipping some patch versions (e.g. `v3.0.1 -> v3.0.3`) are also not affected. We'd then remove the scaffold after a "tolerance period". (How would we determine the length of that tolerance period?)

**Comment by simster7** [comment]
Also perhaps we could include a notice in the release notes the advise users to not upgrade with running workflows. And we can include that notice whenever applicable. It seems to me that we must have already made similar bug fixes that would have broken running nodes in the past.

This could be a good middle ground between users not seeing unexpected, one-time issues during upgraded and us being able to fix bugs that currently exist in the codebase.

**Comment by simster7** [comment]
Decision: add scaffolding code

File: workflow/controller/controller_test.go [hunk 2]
```diff
@@ -301,6 +301,31 @@ func withAnnotation(key, val string) with {
 	return func(pod *apiv1.Pod) { pod.Annotations[key] = val }
 }
 
+// createRunningPods creates the pods that are marked as running in a given test so that they can be accessed by the
+// pod assessor
+func createRunningPods(ctx context.Context, woc *wfOperationCtx, with ...with) {
```
Comment (simster7) [inline workflow/controller/controller_test.go:306]: Useful utility testing function to create pods in the client that are supposed to be running according to the test workflow status.

File: workflow/controller/dag.go [hunk 3]
```diff
@@ -695,7 +695,20 @@ func (d *dagContext) evaluateDependsLogic(taskName string) (bool, bool, error) {
...
+		}
+
+		// Previously we used `depNode.DisplayName` to generate all onExit node names. However, as these can be non-unique
+		// we transitioned to using `depNode.Name` instead, which are guaranteed to be unique. In order to not disrupt
+		// running workflows during upgrade time, we do an additional check to see if there is an onExit node with the old
+		// name (`depNode.DisplayName`).
+		// TODO: This scaffold code should be removed after a couple of "grace period" version upgrades to allow transitions. It was introduced in v3.0.0
+		// See more: https://github.com/argoproj/argo-workflows/issues/5502
+		legacyOnExitNodeName := common.GenerateOnExitNodeName(depNode.DisplayName)
+		if onExitNode := d.wf.GetNodeByName(legacyOnExitNodeName); onExitNode != nil && d.wf.GetNodeByName(depNode.Name).HasChild(onExitNode.ID) {
```
Comment (simster7) [inline workflow/controller/dag.go:711]: This is tested by `TestOnExitDAGStatusCompatibility` in this PR

File: workflow/controller/operator.go [hunk 4]
```diff
@@ -2937,10 +2937,27 @@ func (woc *wfOperationCtx) createTemplateContext(scope wfv1.ResourceScope, resou
...
+		// When the scaffold code is removed, we should only have the following:
+		//
+		// 		onExitNodeName := common.GenerateOnExitNodeName(parentNodeName)
+		//
+		// See more: https://github.com/argoproj/argo-workflows/issues/5502
+		onExitNodeName := common.GenerateOnExitNodeName(parentNodeName)
+		legacyOnExitNodeName := common.GenerateOnExitNodeName(parentNodeDisplayName)
+		if legacyNameNode := woc.wf.GetNodeByName(legacyOnExitNodeName); legacyNameNode != nil && woc.wf.GetNodeByName(parentNodeName).HasChild(legacyNameNode.ID) {
+			onExitNodeName = legacyOnExitNodeName
+		}
```
Comment (simster7) [inline workflow/controller/operator.go:2959]: This is tested by `TestOnExitNameBackwardsCompatibility` in this PR

File: workflow/controller/operator_test.go [hunk 5]
```diff
@@ -6314,3 +6314,223 @@ func TestGenerateOutputResultRegex(t *testing.T) {
 	assert.Equal(t, `steps\.template-name\.outputs\.result`, ref)
 	assert.Equal(t, `steps\[['\"]template-name['\"]\]\.outputs.result`, expr)
 }
+
+const testOnExitNameBackwardsCompatibility = `
+apiVersion: argoproj.io/v1alpha1
```
Comment (simster7) [inline workflow/controller/operator_test.go:6319]: I've already ensured these test workflows are minimal.

**Review by simster7 — COMMENTED** [review]
@alexec Now fully backwards compatible, with tests that ensure it. Ready for another review 🙂

File: workflow/controller/operator.go [hunk 6]
```diff
@@ -2940,7 +2940,24 @@ func (woc *wfOperationCtx) createTemplateContext(scope wfv1.ResourceScope, resou
...
+		// When the scaffold code is removed, we should only have the following:
+		//
+		// 		onExitNodeName := common.GenerateOnExitNodeName(parentNodeName)
+		//
+		// See more: https://github.com/argoproj/argo-workflows/issues/5502
+		onExitNodeName := common.GenerateOnExitNodeName(parentNodeName)
+		legacyOnExitNodeName := common.GenerateOnExitNodeName(parentDisplayName)
+		if legacyNameNode := woc.wf.GetNodeByName(legacyOnExitNodeName); legacyNameNode != nil && woc.wf.GetNodeByName(parentNodeName).HasChild(legacyNameNode.ID) {
+			onExitNodeName = legacyOnExitNodeName
+		}
```
Comment (simster7) [inline workflow/controller/operator.go:2959]: This is tested by `TestOnExitNameBackwardsCompatibility` in this PR

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 7]
```diff
@@ -1806,6 +1806,15 @@ func (n NodeStatus) GetDuration() time.Duration {
 	return n.FinishedAt.Sub(n.StartedAt.Time)
 }
 
+func (n NodeStatus) HasChild(childID string) bool {
+	for _, nodeID := range n.Children {
```
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1810]: unit test please

File: workflow/controller/controller_test.go (hunk 2 above)
Comment (alexec) [reply workflow/controller/controller_test.go:306]: unused? delete?

File: workflow/controller/dag.go (hunk 3 above)
Comment (alexec) [reply workflow/controller/dag.go:711]: nice

File: workflow/controller/operator.go (hunk 6 above)
Comment (alexec) [reply workflow/controller/operator.go:2959]: the block of code looks to be duplicated - maybe create a utility func?

**Review by alexec — CHANGES_REQUESTED** [review]

File: workflow/controller/controller_test.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/controller_test.go:306]: Only the `with` parameter was unused. Removed.

File: workflow/controller/operator.go (hunk 6 above)
Comment (simster7) [reply workflow/controller/operator.go:2959]: I had a utility function before, but once all the parameters were taken into account it didn't really reduce complexity so I decided to keep it like so.

**Review by alexec — APPROVED** [review]

File: pkg/apis/workflow/v1alpha1/workflow_types.go (hunk 7 above)
Comment (simster7) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1810]: Working on it right now, will push soon

**Pull request merged**